File: 0398-random-pick-index/0398-random-pick-index.c



typedef struct {
    int *nums ;
    int size ;
} Solution;


Solution* solutionCreate(int* nums, int numsSize) {       
         Solution *p = (Solution*)malloc(sizeof(Solution)); 
           p->nums = nums ; 
           p->size = numsSize ; 
           return p ;
}

int solutionPick(Solution* obj, int target) {

     int curr = *(obj->nums) -1 ; 
        int count = 1;
        for(int i  = 0 ; i<obj->size ; i++){
            if(obj->nums[i]==target){
                  if(rand()%(count++)==0){
                    curr = i  ;
                  }
            }
        }
        return curr ;
 
}

void solutionFree(Solution* obj) {
    free(obj) ;
}

/**
 * Your Solution struct will be instantiated and called as such:
 * Solution* obj = solutionCreate(nums, numsSize);
 * int param_1 = solutionPick(obj, target);
 
 * solutionFree(obj);
*/

File: 0398-random-pick-index/0398-random-pick-index.java
import java.util.Random ;
class Solution {
     int [] nums ;
     Random r ;
    public Solution(int[] nums) {
        this.nums = nums ;
        this.r = new Random() ;
    }
    
    public int pick(int target) {
        int curr = nums[0] -1 ; 
        int count = 1;
        for(int i  = 0 ; i<nums.length ; i++){
            if(nums[i]==target){
                  if(r.nextInt(count++)==0){
                    curr = i  ;
                  }
            }
        }
        return curr ;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 */